monitor/inclusion_checker.go

```diff
@@ -0,0 +1,195 @@
+	}
+	newHead, entries, err := ic.getEntries(current, int64(sth.TreeSize))
+	if err != nil {
+		inclusionErrors.WithLabelValues("getEntries").Inc()
+		return fmt.Errorf("error retrieving entries from %q: %s", ic.logURI, err)
+	}
+
+	err = ic.checkEntries(certs, entries)
+	if err != nil {
+		inclusionErrors.WithLabelValues("checkEntries").Inc()
+		return fmt.Errorf("error checking retrieved entries for %q: %s", ic.logURI, err)
+	}
+
+	err = ic.db.UpdateIndex(ic.logID, newHead)
+	if err != nil {
+		inclusionErrors.WithLabelValues("updateIndex").Inc()
+		return fmt.Errorf("error updating current index for %q: %s", ic.logURI, err)
+	}
+
+	return nil
+}
+
+func min(a, b int64) int64 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func (ic *inclusionChecker) getEntries(start, end int64) (int64, []ct.LogEntry, error) {
+	if ic.maxGetEntries > 0 && end-start > ic.maxGetEntries {
+		end = start + ic.maxGetEntries
+	}
+	var allEntries []ct.LogEntry
+	for start <= end {
+		batchEnd := min(start+ic.batchSize, end)
+		entries, err := ic.client.GetEntries(context.Background(), start, batchEnd)
+		if err != nil {
+			return 0, nil, err
+		}
+		allEntries = append(allEntries, entries...)
+		start += int64(len(entries))
+	}
+	return start, allEntries, nil
+}
+
+func (ic *inclusionChecker) checkEntries(certs []storage.SubmittedCert, entries []ct.LogEntry) error {
+	// Key structure for our lookup map is as follows: SHA256 hash of the certificate
+	// body concatenated with the byte encoding of the SCT timestamp. This prevents
+	// from having duplicate keys for duplicate submissions with differing SCTs.
+	lookup := make(map[[32]byte]storage.SubmittedCert)
+	for _, cert := range certs {
+		content := make([]byte, len(cert.Cert)+binary.MaxVarintLen64)
+		copy(content, cert.Cert)
+		binary.PutUvarint(content[len(cert.Cert):], cert.Timestamp)
+		lookup[sha256.Sum256(content)] = cert
+	}
+	for _, entry := range entries {
+		var content []byte
+		switch entry.Leaf.TimestampedEntry.EntryType {
+		case ct.X509LogEntryType:
+			content = entry.X509Cert.Raw
+		case ct.PrecertLogEntryType:
+			content = entry.Precert.Submitted.Data
+		}
+		timestampBuf := make([]byte, binary.MaxVarintLen64)
+		binary.PutUvarint(timestampBuf, entry.Leaf.TimestampedEntry.Timestamp)
+		content = append(content, timestampBuf...)
+		h := sha256.Sum256(content)
+		if matching, found := lookup[h]; found {
+			var sct ct.SignedCertificateTimestamp
+			_, err := tls.Unmarshal(matching.SCT, &sct)
+			if err != nil {
+				return fmt.Errorf("error unmarshalling SCT: %s", err)
+			}
+			err = ic.signatureChecker.VerifySCTSignature(sct, entry)
+			if err != nil {
+				return fmt.Errorf("error verifying SCT signature: %s", err)
+			}
+			err = ic.db.MarkCertSeen(matching.ID, ic.clk.Now())
+			if err != nil {
+				return fmt.Errorf("error marking certificate as seen: %s", err)
+			}
+			delete(lookup, h)
+		}
+	}
+
+	var oldest uint64
+	for _, unseen := range lookup {
```

Comment:
How big do we expect `lookup` to get? Is it worth prematurely optimizing for finding the oldest entry faster than `o(n)`?

Reply:
I think it's unlikely to get _very_ large, it should maintain relatively good performance up to quite a big size anyway given the inner logic is a simple comparison.

Reply:
Good point. Def. not sufficient to start doing performance contortions